### PR TITLE
Add CAS method for RawKV

### PR DESCRIPTION
**Pull request by Little-Wallace**

### CODE REVIEW

File: proto/kvrpcpb.proto [hunk 1]
```diff
@@ -1034,3 +1038,22 @@ message RawGetKeyTTLResponse {
...
+    // If this field does not exist, it means that the key-value should only put into database if and only if the key
+    // does not exist.
+    bytes previous_value = 4;
+    string cf = 5;
+    uint64 ttl = 6;
+}
+
+message RawCASResponse {
+    errorpb.Error region_error = 1;
+    string error = 2;
```
Comment (andylokandy) [inline proto/kvrpcpb.proto:1055]: What is this for?

File: proto/kvrpcpb.proto [hunk 2]
```diff
@@ -1034,3 +1038,22 @@ message RawGetKeyTTLResponse {
...
+    bytes value = 3;
+    bool not_equals = 4;
```
Comment (andylokandy) [inline proto/kvrpcpb.proto:1057]: For consistency with L1039

```suggestion
    bool not_equal = 4;
```

File: proto/tikvpb.proto [hunk 3]
```diff
@@ -50,6 +50,9 @@ service Tikv {
     // Get TTL of the key. 
     // If the key is non-TTL, returns 0
     rpc RawGetKeyTTL(kvrpcpb.RawGetKeyTTLRequest) returns (kvrpcpb.RawGetKeyTTLResponse) {}
+    // Put the key-value if and the previous value store in database equal RawCASRequest.previous_value. If not, return
+    // the previous value.
```
Comment (andylokandy) [inline proto/tikvpb.proto:54]:
```suggestion
    // Compare if the value in database equals to `RawCASRequest.previous_value` before putting the new value. If not, the mutation has no effect and the value in database will be returned.
```

**Review by andylokandy — APPROVED** [review]

File: proto/kvrpcpb.proto [hunk 4]
```diff
@@ -446,6 +446,7 @@ message RawPutRequest {
     bytes value = 3;
     string cf = 4;
     uint64 ttl = 5;
+    bool for_cas = 6;
```
Comment (BusyJay) [inline proto/kvrpcpb.proto:449]: How about `check_cas`?

File: proto/kvrpcpb.proto [hunk 5]
```diff
@@ -1034,3 +1038,22 @@ message RawGetKeyTTLResponse {
...
     bool not_found = 4;
 }
+
+message RawCASRequest {
+    Context context = 1;
+    bytes key = 2;
+    bytes value = 3;
+    // If this field does not exist, it means that the key-value should only put into database if and only if the key
+    // does not exist.
+    bytes previous_value = 4;
```
Comment (BusyJay) [inline proto/kvrpcpb.proto:1048]: Actually you can't tell non-exist value from empty value in protobuf.

Comment (Little-Wallace) [reply proto/kvrpcpb.proto:1048]: If the value is a empty string, I think it present non-exist

File: proto/kvrpcpb.proto (hunk 4 above)
Comment (Little-Wallace) [reply proto/kvrpcpb.proto:449]: I think there is no difference between them....

**Review by Connor1996 — APPROVED** [review]

File: proto/kvrpcpb.proto (hunk 5 above)
Comment (andylokandy) [reply proto/kvrpcpb.proto:1048]: Please don't do that. It'll be a legacy problem like this one:

https://github.com/tikv/tikv/issues/6711

Comment (andylokandy) [reply proto/kvrpcpb.proto:1048]: I suggest adding a new bool field `should_not_exist` to indicate that property.

**Review by Connor1996 — COMMENTED** [review]
You need to regenerate the files. There are a lot of changes after you change the version of protoc.

File: proto/kvrpcpb.proto (hunk 4 above)
Comment (BusyJay) [reply proto/kvrpcpb.proto:449]: `for_cas` seems the key is put for CAS purpose. However, the field is meant to be participated in CAS checking.

File: proto/kvrpcpb.proto [hunk 6]
```diff
@@ -1034,3 +1038,23 @@ message RawGetKeyTTLResponse {
     uint64 ttl = 3;
     bool not_found = 4;
 }
+
+message RawCASRequest {
+    Context context = 1;
+    bytes key = 2;
+    bytes value = 3;
+    // If the length of this field equals zero, it means that the key-value should only put into database if and only if the key
```
Comment (BusyJay) [inline proto/kvrpcpb.proto:1046]: Stale comment.

**Pull request merged**